packages/search/provider.tsx

```diff
@@ -0,0 +1,115 @@
+import React, { useMemo, useState, createContext, useContext } from 'react'
+import algoliaSearch, { SearchClient } from 'algoliasearch'
+import searchInsights from 'search-insights'
+import { AlgoliaConfigObject } from './types'
+
+const SearchContext = createContext(null)
+
+interface SearchContextObject {
+  client: SearchClient
+  indexName: string
+  initAlgoliaInsights: () => void
+  isCancelled: boolean
+  logClick: () => void
+  query: string
+  setCancelled: () => void
+  setQuery: () => void
+}
+
+export function useSearch(): SearchContextObject {
+  return useContext(SearchContext)
+}
+
+export default function SearchProvider({
+  children,
+  algoliaConfig = {},
+}: {
+  children: React.ReactNode
+  algoliaConfig: AlgoliaConfigObject
+}): React.ReactElement {
+  const [query, setQuery] = useState('')
+  const [isCancelled, setCancelled] = useState(false)
+
+  const [appId, indexName, apiKey] = [
+    algoliaConfig.appId || process.env.NEXT_PUBLIC_ALGOLIA_APP_ID,
+    algoliaConfig.indexName || process.env.NEXT_PUBLIC_ALGOLIA_INDEX,
+    algoliaConfig.searchOnlyApiKey ||
+      process.env.NEXT_PUBLIC_ALGOLIA_SEARCH_ONLY_API_KEY,
+  ]
+
+  if (!appId || !apiKey || !indexName) {
+    throw new Error(`Missing required Algolia arguments or .env variables for SearchProvider. If using environment variables, ensure the following are present in your environment:
+- NEXT_PUBLIC_ALGOLIA_APP_ID
+- NEXT_PUBLIC_ALGOLIA_SEARCH_ONLY_API_KEY
+- NEXT_PUBLIC_ALGOLIA_INDEX
+If passing the algoliaConfig prop to SearchProvider, ensure the following keys are defined:
+- appId
+- indexName
+- searchOnlyApiKey
+`)
+  }
+
+  const algoliaClient = useMemo(() => algoliaSearch(appId, apiKey), [
+    appId,
+    apiKey,
+  ])
+
+  const client = {
+    search(requests) {
+      return algoliaClient.search(
+        requests.map((request) => {
+          //  instantsearch fires an empty query on page load to ensure results are immediately available
+          //  we exclude that result from our analytics to keep our clickthrough rate clean
+          //  ref: https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/out-of-the-box-analytics/how-to/how-to-remove-empty-search-from-analytics/
+          if (!request.params.query || request.params.query.length === 0) {
+            request.params.analytics = false
+          }
+          return request
+        })
+      )
+    },
+  }
+
+  function initAlgoliaInsights() {
+    searchInsights('init', { appId, apiKey })
+  }
+
+  function logClick(hit) {
+    // --------------
+    // TODO: I was unable to resolve the TypeScript error here.
```

Comment:
I ran into this TypeScript issue, timeboxed myself a few minutes to try to resolve it, and wasn't able to. If anyone has any insight into how to resolve this issue, would love any help 🙇 

I don't love having this messy comment, but I also don't love sinking a ton of time into "fixing" TS errors in code that I'm not really trying to touch.... and given the code here is identical I think perhaps it's acceptable for now? cc @BRKalow the TypeScript wizard 🧙 (also no pressure here, I'm happy to try to dive in further and try to fix this if needed).

Reply:
@zchsh It wasn't you! Looks like the types are wrong from the package. I pushed up a minor change. The types are fixed in v2.1.0, we are currently on v1.8.0. Feel free to take a stab at upgrading if you're feeling _dangerous_.

Reply:
Alright, I ran into what feels like an out-of-scope issue with the upgrade:

Based on the README for v2.x.x, the only breaking change is that `useCookie` will default to `false` rather than `true`. Based on [this note on cookies](https://github.com/algolia/search-insights.js#cookie-usage), and [the `useCookie` entry in the options documentation table](https://github.com/algolia/search-insights.js#2-initialize-the-library), we'll need to explicitly set `useCookie` to `true` in order to track _any_ search events (otherwise they'll error). I think this would affect search events such as `logClick`, [as used in react-search hits](https://github.com/hashicorp/react-components/blob/03e3ed2549d79f632673f702db1a63748063bb95/packages/search/hits.js#L178) and therefore in docs-page, or such as [`logConversion` in learn](https://github.com/hashicorp/learn/blob/4a2458b69844a8eb542baeba008fb5d302794713/lib/hooks/log-search-conversion.js#L15)).

In other words, it seems that as part of the upgrade, and perhaps even as part of our current use, we should consider hooking into `ConsentManager` before calling eg `logClick` or `logConversion`.

So, going to hold off on the upgrade, and have added a task to our web-platform board to discuss 👍